### PR TITLE
Restructure project to use pyproject.toml

### LINKED ISSUE #706: pip install deprecation warning
Hello from a member of the original Paho team 🙂 thanks for continuing the great work!

On macOS Ventura, Python 3.11, pip 23.0

```
 DEPRECATION: paho-mqtt is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
  Running setup.py install for paho-mqtt ... done
```

There are workarounds listed (and the installation works and the library is fine), but Paho probably needs to be updated for the new `pyproject.toml` library metadata format.

### PR DESCRIPTION
**Pull request by semuadmin**
`setup.py` is now [deprecated](https://pip.pypa.io/en/stable/reference/build-system/setup-py/) and is scheduled to be dropped altogether in pip 23.

This PR makes a number of changes to update the project build framework:

1. It updates the paho.mqtt project structure to use `pyproject.toml`, while retaining setuptools as a back-end build platform. 
2. It uses pyproject.toml tool configuration sections were available.
3. It drops support for Python versions <=3.6 as these are now end of life.
4. It adds a minimal VS Code workflow including the following tasks (_Eclipse IDE users can simply ignore these_):
  - The `Build` task will produce wheel (`*.whl`) and sdist (`*.tar.gz`) install packages in the dist folder which can be uploaded to PyPi using twine. 
  - The `Install Locally` task will install this wheel using pip.
  - Other VS Code tasks have been added mirroring the current GitHub Actions (GHA) CI worfklows.

**NB:** _this PR only changes the build setup - no changes have been made to existing application code or test cases, so some existing CI workflows may still fail if they were failing previously_.

Fixes #706

### CODE REVIEW

**Comment by semuadmin** [comment]
I have and Eclipse account and have signed the ECA but the `eclipsefdn/eca` workflow check is still flagged as `errored`. There may be some latency involved in the ECA registration, but if I'm missing a step, please advise. Thanks

**Comment by semuadmin** [comment]
**Note to maintainers:** 

As far as I can tell, this new pyproject.toml build and deployment framework works fine.

But I note that the source code in the repo's master branch has moved on from the source currently deployed in PyPi, though both version numbers are marked as `1.6.1`. If you actually package and install the current master branch, the code doesn't appear to work, while the PyPi version works fine. 

Just a suggestion, but it might be worth keeping the master branch in line with the corresponding release in PyPI and putting any intervening changes in a different 'release candidate' branch? Also, maintaining Releases versions in GitHub can be helpful.

**Comment by cclauss** [comment]
My suggestion would be to replace all the linters (flake8, pylint, pylama, bandit, pyupgrade) with a single, fast (<10 sec) GitHub Action which runs [`ruff`](https://beta.ruff.rs/docs) as in #717

Then a separate GHA can focus on formatting (psf/black) and testing (pytest) and possible add type hints (mypy).

Did you use tools to convert `setup.py` (setuptools-py2cfg) and `setup.cfg` (ini2toml) to `pyproject.toml` or was this a manual conversion?

**Comment by semuadmin** [comment]
> My suggestion would be to replace all the linters (flake8, pylint, pylama, bandit, pyupgrade) with a single, fast (<10 sec) GitHub Action which runs [`ruff`](https://beta.ruff.rs/docs) as in #717

Another day, another linter ;-)

I noticed that the existing project workflow has references to a variety of linters - I'm guessing that may reflect the historical predilections of the individual maintainers. I agree that it would probably be sensible to settle on one, but I'm happy to leave this to the discretion of the maintainers - it's not the main focus of this PR.

Having said that, I can't help but think the maintainers' precious time might be better spent addressing more urgent issues (_like moving off setup.py asap_) than beta-testing a new (_though doubtless very capable_) linter. In the meantime, pylint and flake are mature, proven, capable, robust and natively supported by all popular IDEs, and the ≈3.9 seconds it takes to run pylint against this library on a moderately capable laptop may not be such a big deal.

> Did you use tools to convert setup.py (setuptools-py2cfg) and setup.cfg (ini2toml) to pyproject.toml or was this a manual conversion?

Manual - I already had a template I've used for other Python projects.

File: MANIFEST.in
Comment (cclauss) [inline MANIFEST.in:11]: ⛔ below is because GitHub expects all text files to end with one and only one blank line.

**Review by cclauss — APPROVED** [review]

**Comment by cclauss** [comment]
> I noticed that the existing project workflow has references to a variety of linters

I wrote that before ruff existed.

File: MANIFEST.in
Comment (semuadmin) [reply MANIFEST.in:11]: Hi @cclauss Apologies if I'm missing something here, but I'm not seeing any specific errors or warnings in my build workflow with regard to missing blank lines in MANIFEST.in. Also, if you add blank lines to any json file, black will promptly remove them again! What exactly is the issue? Thanks.

Comment (cclauss) [reply MANIFEST.in:11]: Click the link in my note above and then look just below line 12.

**Comment by semuadmin** [comment]
I've added the blank lines as requested, though I'm still unclear about the necessity? I've run hundreds of GHA workflows with the MANIFEST.in, settings.json and tasks.json files formatted exactly was they are here and I've never seen any warnings or errors in the GHA log (admittedly I'm not using Tox in my GitHub actions).

**Comment by cclauss** [comment]
⛔ https://github.com/eclipse/paho.mqtt.python/pull/712/files#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357R9

**Comment by semuadmin** [comment]
> ⛔ https://github.com/eclipse/paho.mqtt.python/pull/712/files#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357R9

sorry, should be OK now

**Comment by semuadmin** [comment]
Re. the failed checks above, I've not touched the existing GitHub Actions or the source code so these are presumably pre-existing issues:

1) the lint_python step is failing because bandit detects "Use of weak SHA1 hash for security" in client.py. This needs a code fix to use e.g. SHA2.
2) the tox 3.6 check is presumably failing because python 3.6 is now end-of-life and needs to be excluded from the workflow.

Who are the maintainers expecting to resolve these issues?

**Comment by cclauss** [comment]
Please add the failing bandit code to the skip list like:
https://github.com/eclipse/paho.mqtt.python/pull/717/files#diff-14721317565b7c66355f276ec27762054b12f390afd6c35fba85cfaf77d311c7R12

It is a newer code added since that GitHub Action was created.

**Comment by semuadmin** [comment]
Hi @cclauss So I take it you're dropping this PR in favour of your own #719?

File: .github/workflows/lint-python.yml
Comment (cclauss) [inline .github/workflows/lint-python.yml:16]: https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html

File: .github/workflows/lint-python.yml
Comment (cclauss) [inline .github/workflows/lint-python.yml:9]: Please keep dependencies sorted to make it easy to spot missing dependencies and difficult insert duplicates.

Given that build pip setuptools and wheel are used to build and install other dependencies, it is traditional to install that group separately before installing other deps.

Comment (semuadmin) [reply .github/workflows/lint-python.yml]: done - you might want to cancel outdated GHA runs 155-157

File: .github/workflows/lint-python.yml
Comment (cclauss) [inline .github/workflows/lint-python.yml:16]: When typing on the command line it will usually be preferred to use single-letter options but when writing in scripts and similar files, it is preferred to use longer names to make things more self-documenting and less cryptic.

File: .github/workflows/lint-python.yml
Comment (cclauss) [inline .github/workflows/lint-python.yml:17]:
```suggestion
      - run: pip install -editable .  # will require build backend support for PEP660 (setuptools >= 64)
```

File: .github/workflows/lint-python.yml
Comment (cclauss) [inline .github/workflows/lint-python.yml:17]:
```suggestion
      - run: pip install —editable .  # will require build backend support for PEP660 (setuptools >= 64)
```

Comment (semuadmin) [reply .github/workflows/lint-python.yml]: Done

Have also removed requirements*.txt and used relevant sections in pyproject.toml for various pip installs.
I've segregated build, test and code dependencies, but you can cut and dice it any way you prefer.

File: .github/workflows/lint-python.yml
Comment (semuadmin) [reply .github/workflows/lint-python.yml]: Done

File: .github/workflows/lint-python.yml
Comment (cclauss) [inline .github/workflows/lint-python.yml:9]: Does this work?
```suggestion
      - run: pip install .[build,text]
```

Comment (semuadmin) [reply .github/workflows/lint-python.yml:9]: Not sure, but I thought you wanted to segregate build and test deps in your GHA workflows? IMHO it's marginally easier to debug any GHA fails if you have the separate steps.

**Comment by cclauss** [comment]
What happened here?!?  These fixes were vital for getting the test green again.

**Pull request closed without merging**
